src/RA_httpthread.cpp

```diff
@@ -252,31 +247,74 @@ void RAWeb::SetUserAgentString()
-    auto RtlGetVersion = (fnRtlGetVersion)GetProcAddress(GetModuleHandleW(L"ntdll.dll"), "RtlGetVersion");
-    if (RtlGetVersion)
     {
-        RTL_OSVERSIONINFOEXW osVersion = {};
+        // needs to be XP or higher no point in continuing
+        // Windows 7 doesn't have the version helper api....
+        // TBD: Consider Service Pack?
+        {
+            auto IsXPOrGreater =[]() noexcept
+            {
+                OSVERSIONINFOEXW osvi{
+                    sizeof(osvi), HIBYTE(_WIN32_WINNT_WINXP) , LOBYTE(_WIN32_WINNT_WINXP)
+                };
+                const auto dwlConditionMask
+                {
+                    VerSetConditionMask(VerSetConditionMask(VerSetConditionMask(
+                        0ULL, VER_MAJORVERSION, VER_GREATER_EQUAL),
+                        VER_MINORVERSION, VER_GREATER_EQUAL),
+                        VER_SERVICEPACKMAJOR, VER_GREATER_EQUAL)
+                };
+
+                return VerifyVersionInfoW(&osvi, VER_MAJORVERSION | VER_MINORVERSION |
+                                          VER_SERVICEPACKMAJOR, dwlConditionMask) != FALSE;
+            };
+            if (!IsXPOrGreater())
+                return;
+        }
+
+        using fnRtlGetVersion = NTSTATUS(NTAPI*)(PRTL_OSVERSIONINFOEXW lpVersionInformation);
+
+        const auto& ntModule{ ::GetModuleHandleW(L"ntdll.dll") };
+        if (!ntModule)
+            return;
+
+        RTL_OSVERSIONINFOEXW osVersion{};
+        const auto& RtlGetVersion{
+            reinterpret_cast<fnRtlGetVersion>(::GetProcAddress(ntModule, "RtlGetVersion"))
+        };
+        if (!RtlGetVersion)
+            return;
         RtlGetVersion(&osVersion);
-        if (osVersion.dwMajorVersion > 0)
+        
         {
-            sUserAgent.append("WindowsNT ");
-            sUserAgent.append(std::to_string(osVersion.dwMajorVersion));
-            sUserAgent.append(".");
-            sUserAgent.append(std::to_string(osVersion.dwMinorVersion));
+            std::ostringstream oss;
+            oss << "WindowsNT " << osVersion.dwMajorVersion << '.' << osVersion.dwMinorVersion
+                << '.' << osVersion.dwBuildNumber;
+            sUserAgent.append(oss.str());
         }
     }
 
     sUserAgent.append(") Integration/");
+    {
+        std::ostringstream oss;
+        // Compile-time string concatenation will need some user-defined stuff/work.
+        oss << "WindowsNT " << RA_INTEGRATION_VERSION_MAJOR << '.' << RA_INTEGRATION_VERSION_MINOR
+            << '.' << RA_INTEGRATION_VERSION_REVISION << '.' << RA_INTEGRATION_VERSION_MODIFIED;
+        sUserAgent.append(oss.str());
+    }
 
-    auto buffer{ std::make_unique<char[]>(64U) };
-    sprintf_s(buffer.get(), 64U, "%d.%d.%d.%d", RA_INTEGRATION_VERSION_MAJOR, RA_INTEGRATION_VERSION_MINOR,
-              RA_INTEGRATION_VERSION_REVISION, RA_INTEGRATION_VERSION_MODIFIED);
-    sUserAgent.append(buffer.release());
```

Comment:
This is a memory leak. Luckily the method is only called once.
https://github.com/RetroAchievements/RAIntegration/commit/59acb109e85b1f63cdcd5803a503fac8d699effd#diff-5f269daa1d43eae5f0106751651f3a31R274